File: 0031-next-permutation/0031-next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        boolean f = false;
        int index = 0;
        for (int i = nums.length - 1; i >= 1; i--) {
            if (nums[i] > nums[i - 1]) {
                index = i - 1;
                f = true;
                break;
            }
        }

        if (f) {
            int least_max = Integer.MAX_VALUE;
            int index1 = index + 1;
            for (int i = index + 1; i < nums.length; i++) {
                if (nums[i] > nums[index]) {
                    least_max = Math.min(least_max, nums[i]);
                    index1 = (least_max == nums[i]) ? i : index1;
                }
            }
            int temp = nums[index];
            nums[index] = nums[index1];
            nums[index1] = temp;
            Arrays.sort(nums, index + 1, nums.length);
        } else {
            Arrays.sort(nums);
        }
    }
}

File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size();
        int i=-1;
        bool swapped=false;
        for(i=n-1;i>=1;i--){
            if(nums[i-1]<nums[i]){
                int mini=i;
                for(int j=i;j<n;j++)if(nums[j]>nums[i-1]&&nums[j]<nums[mini])mini=j;
                swap(nums[i-1],nums[mini]);
                sort(nums.begin()+i,nums.end());
                swapped=true;
                break;
            }
        }
        if(!swapped)sort(nums.begin(),nums.end());
    }
};